<launch>

    <!-- Some general parameters -->
    <param name="use_sim_time" value="true" />

    <!-- Start Stage simulator with a given environment -->
    <!--        <node name="Stage" pkg="stage_ros" type="stageros" args="$(find continuous_connectivity)/world/cave4_1.world">-->
    <node name="Stage" pkg="stage_ros" type="stageros" args="$(find continuous_connectivity)/world/city8_5.world">
        <param name="base_watchdog_timeout" value="0" />
    </node>
    <node name="R0_MapAlign" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 /map /robot_0/map 100"/>
    <node name="R1_MapAlign" pkg="tf" type="static_transform_publisher" args="100 0 0 0 0 0 /map /robot_1/map 100"/>
    <node name="R2_MapAlign" pkg="tf" type="static_transform_publisher" args="200 0 0 0 0 0 /map /robot_2/map 100"/>
    <node name="R3_MapAlign" pkg="tf" type="static_transform_publisher" args="300 0 0 0 0 0 /map /robot_3/map 100"/>

    <group ns="robot_0">
        <param name="robot_id" value="1" />
        <param name="tf_prefix" type="string" value="robot_0"/>
        <rosparam file="$(find continuous_connectivity)/param/ros.yaml"/>
        <node pkg="nav2d_operator" type="operator" name="operator" output="screen">
            <rosparam file="$(find continuous_connectivity)/param/operator.yaml"/>
            <rosparam file="$(find continuous_connectivity)/param/costmap1.yaml" ns="local_map" />
            <param name="robot_frame" value="base_link"/>
            <param name="odometry_frame" value="odom"/>
        </node>
        <node pkg="nav2d_navigator" type="navigator" name="navigator">
            <rosparam file="$(find continuous_connectivity)/param/navigator.yaml"/>
            <param name="robot_id" value="1"/>
        </node>
        <node pkg="nav2d_navigator" type="explore_client" name="explore_client"/>
        <node name="GetMap" pkg="nav2d_navigator" type="get_map_client" />
        <node name="SetGoal" pkg="nav2d_navigator" type="set_goal_client" />
        <include file="$(find continuous_connectivity)/launch/robot.launch">
            <arg name="robot_name" value="0" />
            <arg name="robot_type" value="2" />
            <arg name="base_stations" value="1,2,3" />
            <arg name="map_width" value="80.0"/>
            <arg name="map_height" value="80.0"/>
        </include>

    </group>

    <group ns="robot_1">
        <param name="robot_id" value="2" />
        <param name="tf_prefix" type="string" value="robot_1"/>
        <rosparam file="$(find continuous_connectivity)/param/ros.yaml"/>
        <node pkg="nav2d_operator" type="operator" name="operator" output="screen">
            <rosparam file="$(find continuous_connectivity)/param/operator.yaml"/>
            <rosparam file="$(find continuous_connectivity)/param/costmap2.yaml" ns="local_map" />
            <param name="robot_frame" value="base_link"/>
            <param name="odometry_frame" value="odom"/>
        </node>
        <node pkg="nav2d_navigator" type="navigator" name="navigator">
            <rosparam file="$(find continuous_connectivity)/param/navigator.yaml"/>
            <param name="robot_id" value="2"/>
        </node>
        <node pkg="nav2d_navigator" type="explore_client" name="explore_client"/>
        <node name="GetMap" pkg="nav2d_navigator" type="get_map_client" />
        <node name="SetGoal" pkg="nav2d_navigator" type="set_goal_client" />

        <include file="$(find continuous_connectivity)/launch/robot.launch">
            <arg name="robot_name" value="1" />
            <arg name="robot_type" value="2" />
            <arg name="base_stations" value="0,2,3" />
            <arg name="map_width" value="80.0"/>
            <arg name="map_height" value="80.0"/>
        </include>
    </group>

    <group ns="robot_2">
        <param name="robot_id" value="3" />
        <param name="tf_prefix" type="string" value="robot_2"/>
        <rosparam file="$(find continuous_connectivity)/param/ros.yaml"/>
        <node pkg="nav2d_operator" type="operator" name="operator" output="screen">
            <rosparam file="$(find continuous_connectivity)/param/operator.yaml"/>
            <rosparam file="$(find continuous_connectivity)/param/costmap3.yaml" ns="local_map" />
            <param name="robot_frame" value="base_link"/>
            <param name="odometry_frame" value="odom"/>
        </node>
        <node pkg="nav2d_navigator" type="navigator" name="navigator">
            <rosparam file="$(find continuous_connectivity)/param/navigator.yaml"/>
            <param name="robot_id" value="3"/>
        </node>
        <node pkg="nav2d_navigator" type="explore_client" name="explore_client"/>
        <node name="GetMap" pkg="nav2d_navigator" type="get_map_client" />
        <node name="SetGoal" pkg="nav2d_navigator" type="set_goal_client" />
        <include file="$(find continuous_connectivity)/launch/robot.launch">
            <arg name="robot_name" value="2" />
            <arg name="robot_type" value="2" />
            <arg name="base_stations" value="0,1,3" />
            <arg name="map_width" value="80.0"/>
            <arg name="map_height" value="80.0"/>
        </include>
    </group>

    <group ns="robot_3">
        <param name="robot_id" value="4" />
        <param name="tf_prefix" type="string" value="robot_3"/>
        <rosparam file="$(find continuous_connectivity)/param/ros.yaml"/>
        <node pkg="nav2d_operator" type="operator" name="operator" output="screen">
            <rosparam file="$(find continuous_connectivity)/param/operator.yaml"/>
            <rosparam file="$(find continuous_connectivity)/param/costmap4.yaml" ns="local_map" />
            <param name="robot_frame" value="base_link"/>
            <param name="odometry_frame" value="odom"/>
        </node>
        <node pkg="nav2d_navigator" type="navigator" name="navigator">
            <rosparam file="$(find continuous_connectivity)/param/navigator.yaml"/>
            <param name="robot_id" value="4"/>
        </node>
        <node pkg="nav2d_navigator" type="explore_client" name="explore_client"/>
        <node name="GetMap" pkg="nav2d_navigator" type="get_map_client" />
        <node name="SetGoal" pkg="nav2d_navigator" type="set_goal_client" />
        <include file="$(find continuous_connectivity)/launch/robot.launch">
            <arg name="robot_name" value="3" />
            <arg name="robot_type" value="2" />
            <arg name="base_stations" value="0,1,2" />
            <arg name="map_width" value="80.0"/>
            <arg name="map_height" value="80.0"/>
        </include>
    </group>

    <!-- RVIZ to view the visualization -->
    <node name="RVIZ" pkg="rviz" type="rviz" args=" -d $(find continuous_connectivity)/param/continuous_connectivity.rviz" output="screen" />

</launch>

